### PR TITLE
Revision range compile functions

### PR DESCRIPTION
**Pull request by boehmseb**
This PR's goal is to make more revisions of our case studies compile by specifying different compile functions for different revision ranges. Therefore, it introduces functionality to check whether a commit lies in a specific _revision range_.

This is currently applied to the following case studies (see se-passau/vara-paper-configs#1):
 
- [ ] doxygen
- [ ]  git
- [x]  gravity
- [ ]  gzip
- [ ]  libvpx
- [ ]  lrzip
- [ ]  tmux
- [ ]  vim
- [x]  x264 (partially)
- [x]  xz (partially)

I suggest to wait with merging until this is done for all the listed case studies.

However, this works mostly for changes in the build system but does not cover all cases of failing builds. 
To cover revisions that cannot be compiled at all, I also work on revision blacklisting (see #150).
For case study-specific dependencies, wee need a system for specifying and checking such dependencies, which will also be a different PR (see se-passau/VaRA#448).

### CODE REVIEW

**Comment by boehmseb** [comment]
If somebody is more familiar with a specific case study, feel free to contribute to this PR. The case study `gravity` can be used as a template for how to use revision ranges.

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/projects/c_projects/xz.py  2
- varats/projects/c_projects/x264.py  2
         

Clones added
============
- varats/projects/c_projects/xz.py  1
- varats/projects/c_projects/gravity.py  1
         

Clones removed
==============
+ varats/projects/c_projects/x264.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4460067&bid=15069528)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/projects/c_projects/xz.py  2
- varats/projects/c_projects/x264.py  2
         

Clones added
============
- varats/projects/c_projects/xz.py  1
- varats/projects/c_projects/gravity.py  1
         

Clones removed
==============
+ varats/projects/c_projects/x264.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4460067&bid=15069528)

File: varats/projects/c_projects/gravity.py [hunk 1]
```diff
@@ -36,8 +37,29 @@ def run_tests(self, runner: run) -> None:
...
 
+        # commit 46133fb47d6da1f0dec27ae23db1d633bc72e9e3 introduced
+        # cmake as build system
+        with local.cwd(self.SRC_FILE):
+            version_id = git("rev-parse", "HEAD").strip()
+            cmake_revisions = get_all_revisions_between(
+                "46133fb47d6da1f0dec27ae23db1d633bc72e9e3", "master")
+
+        if version_id in cmake_revisions:
+            print("CMAKE")
```
Comment (vulder) [inline varats/projects/c_projects/gravity.py]:
```suggestion
```
Remove debug print

File: varats/projects/c_projects/gravity.py [hunk 2]
```diff
@@ -36,8 +37,29 @@ def run_tests(self, runner: run) -> None:
...
+            self.__compile_cmake()
+        else:
+            print("MAKE")
```
Comment (vulder) [inline varats/projects/c_projects/gravity.py]:
```suggestion
```
remove debug print

File: varats/projects/c_projects/xz.py [hunk 3]
```diff
@@ -36,13 +37,24 @@ def run_tests(self, runner: run) -> None:
     def compile(self) -> None:
         self.download()
 
+        # commit 46133fb47d6da1f0dec27ae23db1d633bc72e9e3 introduced
+        # cmake as build system
+        with local.cwd(self.SRC_FILE):
+            version_id = git("rev-parse", "HEAD").strip()
+            cmake_revisions = get_all_revisions_between(
+                "5d018dc03549c1ee4958364712fb0c94e1bf2741", "fda4724d8114fccfa31c1839c15479f350c2fb4c")
```
Comment (vulder) [inline varats/projects/c_projects/xz.py]:
```suggestion
                "5d018dc03549c1ee4958364712fb0c94e1bf2741",
                "fda4724d8114fccfa31c1839c15479f350c2fb4c")
```

File: varats/utils/project_util.py [hunk 4]
```diff
@@ -72,3 +72,16 @@ def get_tagged_commits(project_name: str) -> tp.List[tp.Tuple[str, str]]:
             for ref_split in [ref.strip().split() for ref in ref_list]
         ]
         return refs
+
+
+def get_all_revisions_between(a: str, b: str) -> tp.List[str]:
```
Comment (vulder) [inline varats/utils/project_util.py]:
```suggestion
def get_all_revisions_between(c_start: str, c_end: str) -> tp.List[str]:
```
var names could be a bit more precise :smile: `c_start` or `start_commit`... etc

**Review by vulder — CHANGES_REQUESTED** [review]

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/projects/c_projects/xz.py  2
- varats/projects/c_projects/x264.py  2
         

Clones added
============
- varats/projects/c_projects/xz.py  1
- varats/projects/c_projects/gravity.py  1
         

Clones removed
==============
+ varats/projects/c_projects/x264.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4460067&bid=15069528)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/projects/c_projects/xz.py  2
- varats/projects/c_projects/x264.py  2
         

Clones added
============
- varats/projects/c_projects/xz.py  1
- varats/projects/c_projects/gravity.py  1
         

Clones removed
==============
+ varats/projects/c_projects/x264.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4460067&bid=15069528)

**Review by vulder — APPROVED** [review]
lgtm

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/projects/c_projects/xz.py  2
- varats/projects/c_projects/x264.py  2
         

Clones added
============
- varats/projects/c_projects/xz.py  2
- varats/projects/c_projects/gravity.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=4460067&bid=15069528)

**Pull request merged**